### PR TITLE
Phase 6: SWAD — Stochastic Weight Averaging Densely

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

SWAD (Cha et al., NeurIPS 2021) averages model checkpoints densely from within a flat loss region, landing in a flatter basin with better generalization. The `--swad` flag is already implemented in train.py but has never been tested. SWAD is mechanistically distinct from snapshot ensemble (failed PR #1807, which averaged across widely-separated cyclical restarts). SWAD averages within a single stable plateau.

The flat-basin argument aligns with why Lion (sign-based, low-curvature updates) works well — we are already in a geometry where SWAD should thrive.

**Key difference from EMA:** EMA applies exponential decay to recent parameters. SWAD collects checkpoints during a validation-stable window and takes their uniform average — targeting the flat center of a loss basin rather than tracking the latest parameters.

### Code changes needed

**CRITICAL BUG FIX: SWAD does not average the surface refinement head.** The SWAD code in train.py (lines 2147-2167) only collects and averages `_base_model.state_dict()`. The `refine_head` is a SEPARATE module and is NOT included. You MUST fix this:

1. In the SWAD checkpoint collection (around line 2163), also collect refine_head state:
```python
snap = {k: v.cpu().clone() for k, v in _base_model.state_dict().items()}
refine_snap = {k: v.cpu().clone() for k, v in refine_head.state_dict().items()} if refine_head is not None else None
swad_checkpoints.append((snap, refine_snap))
```

2. In the SWAD averaging (around line 2156-2161), also average and load refine_head:
```python
avg_state = {k: torch.stack([c[0][k].float() for c in swad_checkpoints]).mean(0).to(device)
             for k in swad_checkpoints[0][0]}
if ema_model is None:
    ema_model = deepcopy(_base_model)
ema_model.load_state_dict(avg_state)

# Also average refine head
if refine_head is not None and swad_checkpoints[0][1] is not None:
    avg_refine = {k: torch.stack([c[1][k].float() for c in swad_checkpoints]).mean(0).to(device)
                  for k in swad_checkpoints[0][1]}
    if ema_refine_head is None:
        ema_refine_head = deepcopy(refine_head._orig_mod if hasattr(refine_head, '_orig_mod') else refine_head)
    ema_refine_head.load_state_dict(avg_refine)
```

3. Update the `len(swad_checkpoints) > 20` check to work with tuples.

### Experiment matrix (8 GPUs)

| GPU | Config | Seed |
|-----|--------|------|
| 0 | Baseline (no SWAD) | 42 |
| 1 | Baseline (no SWAD) | 43 |
| 2 | `--swad` | 42 |
| 3 | `--swad` | 43 |
| 4 | `--swad` | 44 |
| 5 | `--swad` | 45 |
| 6 | `--swad` | 66 |
| 7 | `--swad` | 67 |

Use `--wandb_group phase6/swad-averaging` for all runs.

**Base command (SWAD):**
```bash
python train.py --agent edward --wandb_group phase6/swad-averaging \
  --wandb_name "edward/swad-s${seed}" \
  --swad \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --disable_pcgrad --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --seed ${seed}
```

**Base command (baseline, no SWAD):**
Same as above but WITHOUT `--swad`.

### Important notes
- DO NOT modify MAX_EPOCHS (500) or MAX_TIMEOUT (180.0) in train.py
- SWAD suppresses EMA updates — this is expected behavior (line 1805)
- SWAD collection triggers when val_loss < 50% of initial val_loss
- SWAD averaging triggers when val_loss first goes UP during collection
- The sliding window is capped at 20 checkpoints

## Baseline

Current best (8-seed ensemble, PR #2080):
- p_in: **12.2** | p_oodc: **6.7** | p_tan: **29.1** | p_re: **5.8**

Single-model baseline (8-seed mean):
- p_in: **13.03** | p_oodc: **7.83** | p_tan: **30.29** | p_re: **6.45**

Compare SWAD single-model results against the SINGLE-MODEL baseline (not the ensemble).

## Reporting

Post a comment with results in table format including:
- Individual run metrics (p_in, p_oodc, p_tan, p_re) for all 8 runs
- Mean ± std for SWAD runs vs baseline runs
- W&B run IDs

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT edward:

## Results

### Bug Fix Applied
Fixed SWAD to also collect and average the surface refinement head (`refine_head`). Previously SWAD only averaged `_base_model.state_dict()`, completely missing the separate `refine_head` module.

### Individual Run Results

| Run | Config | Seed | p_in | p_oodc | p_tan | p_re | W&B Run |
|-----|--------|------|------|--------|-------|------|---------|
| 1 | Baseline | 42 | **13.2** | **8.0** | **29.7** | **6.4** | 3mim1mhi |
| 2 | Baseline | 43 | **12.8** | **8.2** | **30.3** | **6.4** | gk16mcse |
| 3 | SWAD | 42 | 49.4 | 37.2 | 57.2 | 26.9 | 1ee50z25 |
| 4 | SWAD | 43 | 54.2 | 41.2 | 64.9 | 29.4 | hbm6rfcg |
| 5 | SWAD | 44 | 46.5 | 38.9 | 55.4 | 27.9 | hsvhokae |
| 6 | SWAD | 45 | 43.9 | 35.0 | 55.3 | 24.6 | r632qi5f |
| 7 | SWAD | 66 | 45.8 | 36.3 | 58.3 | 26.3 | 86sd67n7 |
| 8 | SWAD | 67 | 47.9 | 39.6 | 58.9 | 31.5 | tm0513wp |

### Summary Statistics

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| Baseline mean (n=2) | **13.0** | **8.1** | **30.0** | **6.4** |
| SWAD mean (n=6) | 47.9 | 38.0 | 58.3 | 27.8 |
| **SWAD vs Baseline** | **+268%** | **+369%** | **+94%** | **+334%** |

### Command Used
```bash
# Baseline
CUDA_VISIBLE_DEVICES=0 python train.py --agent edward \
  --wandb_group phase6/swad-averaging --wandb_name "edward/baseline-s42" \
  --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed 42

# SWAD
CUDA_VISIBLE_DEVICES=2 python train.py --agent edward \
  --wandb_group phase6/swad-averaging --wandb_name "edward/swad-s42" \
  --swad \
  --asinh_pressure --asinh_scale 0.75 [all baseline flags] --seed 42
```

W&B group: `phase6/swad-averaging`

### What happened

**Catastrophic failure.** SWAD is 3-4x worse than baseline across ALL metrics and ALL seeds.

**Root cause:** SWAD suppresses EMA updates (line 1805: `not cfg.swad`). SWAD collection only begins when `val_loss < 0.5 * initial_val_loss`. In our setup with only 1 val evaluation at the very end (after timeout), SWAD never gets a chance to collect OR average checkpoints. The model runs the entire training without EMA smoothing, and the final raw-weight model is dramatically worse than the EMA-smoothed version.

This is a fundamental incompatibility: SWAD needs frequent val evaluations during training to detect the "flat basin" window, but our training loop only evaluates at the end. Without intermediate val checks, SWAD can never trigger, and by disabling EMA it removes the only mechanism that produces a good final model.

**Key insight:** EMA is load-bearing in this architecture. Any technique that disables EMA (SWAD, SWA) will catastrophically fail unless it provides an equivalent or better weight averaging mechanism that actually activates.

### Suggested follow-ups
- SWAD could work if val evaluation happened every N epochs (e.g., every 10 epochs), allowing the collection window to trigger. But this would add significant training time.
- A simpler approach: keep EMA active and use SWAD as an *additional* averaging on top, rather than a replacement. This would require modifying the SWAD code to not suppress EMA.
- Given the confirmed dead ends list already includes NOBLE and physics losses, this confirms the pattern: EMA is essential and non-negotiable in this training setup.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2094 because Closing — catastrophic failure. SWAD is fundamentally incompatible with our training loop.

| Config | p_in | p_oodc | p_tan | p_re |
|--------|------|--------|-------|------|
| Baseline | 13.0 | 8.1 | 30.0 | 6.4 |
| SWAD (6-seed mean) | 47.9 | 38.0 | 58.3 | 27.8 |
| **Δ** | **+268%** | **+369%** | **+94%** | **+334%** |

Root cause: SWAD suppresses EMA updates (line 1805) but requires intermediate val evaluations to trigger its checkpoint collection window. Our loop evaluates val only at the end → SWAD never activates → model trains without EMA → raw weights are catastrophically worse.

Key insight: **EMA is load-bearing in this architecture.** Any technique that disables EMA without providing an equivalent smoothing mechanism will fail catastrophically. Good bug fix on the refine_head inclusion, but the fundamental incompatibility is the eval schedule, not the head averaging. Student's analysis of the root cause is correct..
